src/repository.rs

```diff
@@ -19,53 +23,142 @@ pub use self::http::{HttpRepository, HttpRepositoryBuilder};
 mod ephemeral;
 pub use self::ephemeral::EphemeralRepository;
 
-/// Top-level trait that represents a TUF repository and contains all the ways it can be interacted
-/// with.
-pub trait Repository<D>
-where
-    D: DataInterchange + Sync,
-{
-    /// Store signed metadata.
+/// A readable TUF repository.
+pub trait RepositoryProvider {
```

Comment:
Why did you move the `DataInterchange` parameter onto the methods? This has the side effect of us not being able to create a `Box<dyn RepositoryProvider>`.

I suppose this would allow us to interact with a repository that supports interchanges, but I'm not sure if that's worth trying to support just yet, since we (and the tuf spec) hasn't actually figured out how alternative interchanges would work.

Reply:
Because `Repository<EphemeralRepository<Json>, Json>` was redundant, and the repository provider is where the metadata and targets are, in whatever form they take.  Yeah, it is a bit weird to allow a provider to provide different data interchanges when we don't have a story for what that looks like yet.

I'll plumb the type parameter back through the trait itself.